### DISCUSSION TITLE
테스트코드 메서드명에 관하여

### DISCUSSION THREAD

**Discussion by yybmion** [Attica-org/athens-backend#139] [General]
안녕하세요 **Athens** 여러분 
다들 테스트 코드를 열심히 작성하고 계실텐데요.
테스트코드의 메서드 명에 대해 함께 의논해 보고 통일하는 것이 좋겠다고 판단하여 discussion 남깁니다.

다음은 7가지 인기 단위 테스트 네이밍 전략이라고 합니다.
https://it-is-mine.tistory.com/3

해당 블로그의 **7번 방법**인 

```markdown
**given_사전조건_When_테스트상태_Expect_기대결과**
```

방법은 직관적으로 테스트의 동작을 알아볼 수 있다는 장점이 있지만 메소드 명이 너무 길어진다는 단점이 있다고 생각합니다. 

따라서 저의 의견은 **테스트 동작은 명확하게 전달하되 메서드 명을 조금이라도 줄이기 위해** **5번 방법**인

```markdown
**should_기대결과_When_테스트상태**  방법을 제시합니다. 
```

다른 분들의 의견은 어떤가요?
@big-cir @Starlight258

**Comment by big-cir** [comment]
저는 해당 링크의 3, 4번을 선호하는 편입니다.
메소드 내부에서 given, when, then을 나눠서 작성하기 때문에 메소드 네이밍까지 명시 할 필요성을 느끼지는 못했습니다.

**Reply by yybmion** [reply]
테스트할 기능만 간단하게 메소드에 명시하는 방법도 좋아보이네요!
의견 감사합니다 👍

**Comment by Starlight258** [comment]
저는 이전 pr에서 테스트 작성시에 `given_주어진상황when_테스트할행동_then_예상결과` 로 작성했었어요.
그런데 `@DisplayName`에서 충분히 상황,행동,조건을 명시하기도 하고, @big-cir 님 말씀처럼 메서드 내부에서` given,when,then`을 나누어 작성하기도 해서 메서드도 굳이 길게 영어로 풀어써야하나라는 생각이 들었어요.

@yybmion 이 제시한 방법처럼 `should_기대결과_When_테스트상태` 으로 간결하게 작성하는 것도 좋아보이네요 👍
그런데 개인적으로 기대 결과(then)이 테스트 상태(when) 먼저 명시되어있어서 테스트 코드 작성 흐름과 반대되는 부분이 걱정이네요.

다른 프로젝트에서는 어떻게 하는지 궁금하네요,, 아예 간단하게 작성하되 테스트 메서드명이 일관성있게 그룹화되면 좋을텐데요 🤔
좋은 디스커션 주제네요 😃👍

**Reply by Starlight258** [reply]
링크에서는 저도 4번이 제일 좋은 것 같아요. 
기능에 대해 간단히 작성하는 것이 테스트 코드 작성 비용이 줄어들 것 같아요. 😀

그리고 만약 테스트 메서드명을 간단하게 작성하게 된다면, 최소한의 틀은 정하는 것이 어떨까요? (블로그 예시의 형식은 일관적이지는 않네요)
테스트 메서드 명도 겹치지 않으면서 해당 테스트가 어떤 테스트인지에 대한 맥락이 잘 제공될 것 같아요.

예를 들어 4번을 채택한다면, `성공/실패 + 테스트할 기능 + 조건` 형식을 제안드려요.
`success_createChat_withValidParameters()`
`failure_createChat_withNonexistentAgoraId()`

테스트 명이 겹치지 않으면서 일관성있게 작성할 수 있을 것 같아요.

추가로 테스트명을 한글로 작성하는 것에 대해 어떠신지 궁금합니다. 😀
`성공_채팅방생성_유효한파라미터사용()`
`실패_채팅방생성_존재하지않는AgoraId사용()`

테스트 작성시에 시간이 덜 걸리기도하고, 빠르게 파악이 가능할 것 같아요.

> given_when_then_보다는 짧긴한데, 생각해보니 해당 형식도 짧은 것은 아니긴하네요! 😅

**Reply by big-cir** [reply]
기존에 작성해 주신 것처럼 이름에 given, when, then으로 명시하는 것이 어떤 흐름으로 동작할지 예측할 수 있다는 장점도 느껴졌었습니다.
영어로 읽는 것에 큰 무리는 없지만, 테스트 양이 많아지면 영어 울렁증이 있는 저로서는 아무래도 한글보다는 에너지(?)를 더 필요로 하지 않을까 생각했습니다.. ㅋㅋㅋㅋㅋ 한글로도 작성하는 것도 좋아 보이네요!

한글로 작성할 때도 작성하는 스타일이 많이 다른것 같아서 이걸로 채택하게 된다면 컨벤션이 필요할 것 같습니다!
```
ex)
1. 성공_채팅방_생성_유효한_파라미터_사용
2. 성공_채팅방생성_유효한파라미터_사용
```
이런식으로 언더바(_)를 어디에서 선언할 지 많이 다른것 같습니다.

**Reply by Starlight258** [reply]
저도 한글이 좀 더 읽기 좋고 작성하기에도 편해서 좋습니다 😃
그렇군요 한글로 하게 된다면 언더바를 띄어쓰기 기준으로 할지, 형식을 나누는 기준으로 할지 등등을 정할 필요가 있네요! 
전 형식을 기준으로 _를 작성하는 것이 좋은 것 같아요. _를 형식 기준으로 나누면 끊어서 빠르게 읽을 수 있어서요 👍
`성공_채팅방생성_유효한파라미터사용`

**Reply by yybmion** [reply]
그렇군요 제안해주신 방법도 좋아보이네요 👍 
`성공/실패 + 테스트할 기능 + 조건` 
저 또한 민트님 의견과 같이 `성공_채팅방생성_유효한파라미터사용` 다음과 같은 형식으로 작성하면 좋을 것 같네요.
